pkg/operation/botanist/component/etcd/etcd.go

```diff
@@ -265,20 +279,83 @@ func (e *etcd) Deploy(ctx context.Context) error {
 		return err
 	}
 
-	if _, err := controllerutils.GetAndCreateOrMergePatch(ctx, e.client, networkPolicy, func() error {
-		networkPolicy.Annotations = map[string]string{
+	// add peer certs if shoot has HA control plane
+	var (
+		etcdPeerCASecret *corev1.Secret
+		peerServerSecret *corev1.Secret
+	)
+	if _, ok := e.annotations[v1beta1constants.ShootAlphaControlPlaneHighAvailability]; ok && gardenletfeatures.FeatureGate.Enabled(features.HAControlPlanes) {
+		etcdPeerCASecret, found = e.secretsManager.Get(v1beta1constants.SecretNameCAETCDPeer)
+		if !found {
+			return fmt.Errorf("secret %q not found", v1beta1constants.SecretNameCAETCDPeer)
+		}
+
+		peerServerSecret, err = e.secretsManager.Generate(ctx, &secretutils.CertificateSecretConfig{
+			Name:                        secretNamePrefixPeerServer + e.role,
+			CommonName:                  "etcd-server",
+			DNSNames:                    e.serviceDNSNames(),
+			CertType:                    secretutils.ServerClientCert,
+			SkipPublishingCACertificate: true,
+		}, secretsmanager.SignedByCA(v1beta1constants.SecretNameCAETCDPeer), secretsmanager.Rotate(secretsmanager.InPlace))
+		if err != nil {
+			return err
+		}
+	}
+
+	// add pod anti-affinity rules for etcd if shoot has a HA control plane
+	if annotationValue, ok := e.annotations[v1beta1constants.ShootAlphaControlPlaneHighAvailability]; ok && gardenletfeatures.FeatureGate.Enabled(features.HAControlPlanes) {
+		if annotationValue == v1beta1constants.ShootAlphaControlPlaneHighAvailabilityMultiZonal {
+			schedulingConstraints.Affinity = &corev1.Affinity{
+				PodAntiAffinity: &corev1.PodAntiAffinity{
+					RequiredDuringSchedulingIgnoredDuringExecution: []corev1.PodAffinityTerm{
+						{
+							TopologyKey: "topology.kubernetes.io/zone",
+							LabelSelector: &metav1.LabelSelector{
+								MatchLabels: e.getPodLabels(),
+							},
+						},
+					},
+				},
+			}
+		} else {
+			schedulingConstraints.Affinity = &corev1.Affinity{
+				PodAntiAffinity: &corev1.PodAntiAffinity{
+					RequiredDuringSchedulingIgnoredDuringExecution: []corev1.PodAffinityTerm{
+						{
+							TopologyKey: "kubernetes.io/hostname",
+							LabelSelector: &metav1.LabelSelector{
+								MatchLabels: e.getPodLabels(),
+							},
+						},
+					},
+					PreferredDuringSchedulingIgnoredDuringExecution: []corev1.WeightedPodAffinityTerm{
```

Comment:
if `multi-zonal` is not set then is it required to add another weighted affinity rule based on topology-key= `topology.kubernetes.io/zone`? Would the first pod anti-affinity rule be sufficient?
Also in `scheduler/reconciler.go` in function `filterMultiZonalSeeds` you are currently only returning non-multizonal seeds for this condition so would you still need this additional affinity?

Reply:
Right now, the logic is:
If HA and multi-zonal, `required pod anti-affinity across zones`
If HA and not multi-zonal, `required pod anti-affinity across nodes` and `preferred pod anti-affinity across zones`.

I had written this logic before the subsequent discussions about confining the etcd members to a single zone if MAZ seeds are not present. So does it make sense to remove the `preferred pod anti-affinity across zones`, or rather change it to `required/preferred pod affinity within a zone` (to confine all members to a single zone to save on network costs)?

> Also in scheduler/reconciler.go in function filterMultiZonalSeeds you are currently only returning non-multizonal seeds for this condition so would you still need this additional affinity?
There is still a chance that a shoot can be pinned to a seed using `seedName` field in the shoot spec. So the gardenlet cannot safely make any assumptions about what kind of seed the shoot is hosted on. So it's better that the all necessary affinity rules are defined here.